File: 3.c
#include <math.h>
#include <stdio.h>

float coords[2][2] = {
    {40.7128, -74.0060}, // New York City
    {42.3601, -71.0589}, // Boston
};

#define PI 3.14159265358979323846
#define EARTH_RADIUS 6371000 // in meters

// Convert degrees to radians
float to_radians(float degrees) { return degrees * (PI / 180.0); }

// Haversine formula to calculate distance
inline float haversine(float lat1, float lon1, float lat2, float lon2) {

  // Convert from degrees to radians
  float lat1_rad = to_radians(lat1);
  float lon1_rad = to_radians(lon1);
  float lat2_rad = to_radians(lat2);
  float lon2_rad = to_radians(lon2);

  // Difference in coordinates
  float delta_lat = lat2_rad - lat1_rad;
  float delta_lon = lon2_rad - lon1_rad;

  // Haversine formula
  float a = sinf(delta_lat / 2) * sinf(delta_lat / 2) +
            cosf(lat1_rad) * cosf(lat2_rad) * sinf(delta_lon / 2) *
                sinf(delta_lon / 2);

  float c = 2 * asinf(sqrtf(a));
  return EARTH_RADIUS * c;
}

inline float haversine_inequality(float lat1, float lon1, float lat2,
                                  float lon2) {

  // Convert from degrees to radians
  float lat1_rad = to_radians(lat1);
  float lon1_rad = to_radians(lon1);
  float lat2_rad = to_radians(lat2);
  float lon2_rad = to_radians(lon2);

  // Difference in coordinates
  float delta_lat = lat2_rad - lat1_rad;
  float delta_lon = lon2_rad - lon1_rad;

  // Haversine formula
  float a = sinf(delta_lat / 2) * sinf(delta_lat / 2) +
            cosf(lat1_rad) * cosf(lat2_rad) * sinf(delta_lon / 2) *
                sinf(delta_lon / 2);

  return a;
}

int main() {
  float distance;
  for (int it = 0; it < 100000000; it++) {
    float lat1 = coords[it % 2][0];
    float lon1 = coords[it % 2][1];
    float lat2 = coords[(it + 1) % 2][0];
    float lon2 = coords[(it + 1) % 2][1];
    distance = haversine(lat1, lon1, lat2, lon2);
    asm volatile("" : "+r"(distance)::);
  }

  printf("Distance: %f meters\n", distance);
  return 0;
}

File: 3.py
import math
from numba import jit

EARTH_RADIUS = 6371000  # Earth radius in meters


@jit(nopython=True)
def to_radians(degrees):
    return math.radians(degrees)


@jit(nopython=True)
def haversine_distance(lat1, lon1, lat2, lon2):
    # Convert from degrees to radians
    lat1_rad = to_radians(lat1)
    lon1_rad = to_radians(lon1)
    lat2_rad = to_radians(lat2)
    lon2_rad = to_radians(lon2)

    # Difference in coordinates
    delta_lat = lat2_rad - lat1_rad
    delta_lon = lon2_rad - lon1_rad

    # Modified Haversine formula
    delta_lat_factor = math.sin(delta_lat / 2) ** 2
    mean_lat_factor = math.sin((lat1_rad + lat2_rad) / 2) ** 2
    delta_lon_factor = math.sin(delta_lon / 2) ** 2

    a = delta_lat_factor + (1 - delta_lat_factor - mean_lat_factor) * delta_lon_factor
    c = 2 * math.asin(math.sqrt(a))
    return EARTH_RADIUS * c


# Example usage
lat1 = 40.7128  # New York City latitude
lon1 = -74.0060  # New York City longitude
lat2 = 42.3601  # Boston latitude
lon2 = -71.0589  # Boston longitude

# Compute the distance
for i in range(10_000_000):
    distance = haversine_distance(lat1, lon1, lat2, lon2)
print(f"Distance: {distance} meters")
